js/node_shared/unix.js

```diff
@@ -0,0 +1,369 @@
+  var r = caml_unix_lstat(name);
+  r[9] = caml_int64_of_int32(r[9]);
+  return r;
+}
+
+//Provides: unix_realpath
+function unix_realpath(path) {
+  return globalThis.fs.realpathSync(path);
+}
+
+//Provides: unix_umask
+function unix_umask(newmask) {
+  let oldmask = globalThis.process.umask(newmask);
+  return oldmask;
+}
+
+//Provides: unix_write
+//Requires: caml_sys_fds
+function unix_write(fd, buf, ofs, len) {
+  console.error("unix_write");
+  console.error(fd);
+  console.error(ofs);
+  console.error(len);
+
+  let file = caml_sys_fds[fd];
+
+  // The reason why this gymnastics is needed, instead of just calling file.write
+  // on the buffer buf.c, is because for some reason the call to writeSync is
+  // copying all of the bytes of the buffer, instead of the requested length from
+  // the offset.
+  // I observed previously that, for a 1024 byte buffer, it would copy all 1024
+  // bytes to the target, instead of the requested 178.
+  // So we just construct a new buffer of what we really want, to ensure the call
+  // gets it right.
+  let bufferSlice = buf.c.slice(ofs, ofs + len);
+
+  // The actual length of our buffer, which depends upon whether the index of
+  // ofs + len is greater than the buffer that we are reading from.
+  let bufferSliceLength =
+    ofs + len < bufferSlice.length ? ofs + len : bufferSlice.length;
+
+  // The implementation of MlNodeFd.write is such that
+  // this always returns 0 if this succeeds.
+  // So let's ignore it.
+  let bytesWritten = file.write(null, bufferSlice, 0, bufferSliceLength);
+
+  return bufferSliceLength;
+}
+
+//Provides: unix_read
+//Requires: caml_sys_fds, caml_convert_bytes_to_array
+function unix_read(fd, buf, ofs, len) {
+  // buf is of type `bytes` in OCaml, which in JSCaml is
+  // represented by an MlBytes object
+  console.error("unix_read");
+  console.error(fd);
+  console.error(ofs);
+  console.error(len);
+
+  // This will ensure buf.c is an array.
+  // We only need to do this if it's not already an array, though.
+  if (buf.t != 4) caml_convert_bytes_to_array(buf);
+
+  const UNIX_BUFFER_SIZE = 65536;
+  let length = len <= UNIX_BUFFER_SIZE ? len : UNIX_BUFFER_SIZE;
+
+  // We do the things we do here because we want a new buffer of the
+  // truncated contents of `buf`, up to the UNIX_BUFFER_SIZE.
+  // We will read the output we are trying to consume into this buffer,
+  // then copy it back to the MlBytes object.
+  const buffer = Buffer.alloc(length);
+  const from_buffer = Buffer.from(buf.c);
+  from_buffer.copy(buffer, 0, 0, length);
+
+  if (fd >= 20000) {
+    // Via our protocol, this is a unique identifier associated to the
+    // pipe (and by extension, the spawned process), which is mapped to
+    // the corresponding child.
+    // Because the child was spawned via 'pipe' stdio, we should be able
+    // to just read the child's output fd to get the data.
+    if (!globalThis.fdTable.has(fd)) {
+      console.error(`fd ${fd} not in fdTable`);
```

Comment:
maybe failwith here?

Reply:
This is actually not a failure case, this might happen. What happens is that `Bos` will call `read` repeatedly until it exits. Because we synchronously read, we actually try to get all the data within a single read, so the first read serves to just return to us the data -- the second read returns a `0`, so that the `Bos` code knows that its done reading.

I'll add a comment.